internal/pkg/walker.go

```diff
@@ -0,0 +1,148 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkg
+
+import (
+	"bytes"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/errors"
+	kptfilev1alpha2 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1alpha2"
+	"github.com/monochromegane/go-gitignore"
+	"sigs.k8s.io/kustomize/kyaml/copyutil"
+)
+
+// AllFileMatchFunc is a matcher func that matches all files.
+var AllFileMatchFunc FileMatchFunc = func(string, os.FileInfo) bool {
+	return true
+}
+
+// YamlFileMatchFunc is a matcher func that matches all files with .yaml or
+// .yml extensions.
+var YamlFileMatchFunc FileMatchFunc = func(_ string, info os.FileInfo) bool {
+	if info.IsDir() {
+		return true
+	}
+
+	ext := filepath.Ext(info.Name())
+	if ext == ".yaml" || ext == ".yml" {
+		return true
+	}
+	return false
+}
+
+// KptfileYamlFileMatchFunc is a matcher func that matches the Kptfile and files
+// with either .yaml or .yml extensions.
+var KptfileYamlFileMatchFunc FileMatchFunc = func(s string, info os.FileInfo) bool {
+	if info.IsDir() {
+		return true
+	}
+
+	if info.Name() == kptfilev1alpha2.KptFileName {
+		return true
+	}
+
+	ext := filepath.Ext(info.Name())
+	if ext == ".yaml" || ext == ".yml" {
+		return true
+	}
+	return false
+}
+
+// WalkFunc defines the type for the callback function used by the
+// Walker.Walk function.
+type WalkFunc func(string, os.FileInfo, error) error
+
+// FileMatchFunc is the type of a function that can be provided to the
+// Walker to filter the files that will be provided to the caller.
+type FileMatchFunc func(string, os.FileInfo) bool
+
+// Walker provides functionality for walking the content of a
+// kpt package. It allows for customization of the files and directories
+// that will be provided to the caller through the callback function.
+type Walker struct {
+	// FileMatchFunc defines the function that will be used by the walker to
+	// filter which files should be provided to the callback. By default, all
+	// files will be provided.
+	FileMatchFunc FileMatchFunc
+
+	// HonorKptfileIgnore defines whether the walker should honor the ignore
+	// list from the package Kptfile.
+	HonorKptfileIgnore bool
+}
+
+// Walk traverses the kpt package provided and invokes the provided
+// callback function for every file and directory in the package that
+// matches the criteria defined in the Walker.
+func (p *Walker) Walk(pkg *Pkg, cb WalkFunc) error {
+	kf, err := pkg.Kptfile()
+
+	// Any other errors than os.ErrNotExist means we should error out.
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+
+	// If the Kptfile could be read and it has at least one ignore pattern,
+	// turn it into a string.
+	var ignorePatterns string
+	if err == nil && len(kf.Ignore) > 0 {
+		ignorePatterns = strings.Join(kf.Ignore, "\n")
+	}
+
+	pkgPath := pkg.UniquePath.String()
```

Comment:
We should prefer `string(pkg.UniquePath)` over `pkg.UniquePath.String()` because `.String()` may not always return what we want here simply because it can be changed for pretty display.